crates/shared/src/bad_token/token_owner_finder/ethplorer.rs

```diff
@@ -0,0 +1,167 @@
+use super::TokenOwnerProposing;
+use crate::rate_limiter::{back_off, RateLimiter, RateLimitingStrategy};
+use anyhow::{ensure, Result};
+use ethcontract::H160;
+use prometheus::IntCounterVec;
+use prometheus_metric_storage::MetricStorage;
+use reqwest::{Client, StatusCode, Url};
+use serde::Deserialize;
+
+const BASE: &str = "https://api.ethplorer.io/getTopTokenHolders/";
+const FREE_API_KEY: &str = "freekey";
+
+pub struct EthplorerTokenOwnerFinder {
+    client: Client,
+    base: Url,
+    api_key: String,
+
+    /// The low tiers for Ethplorer have very aggressive rate limiting, so be sure to setup a rate
+    /// limiter for Ethplorer requests.
+    rate_limiter: Option<RateLimiter>,
+}
+
+impl EthplorerTokenOwnerFinder {
+    pub fn try_with_network(
+        client: Client,
+        api_key: Option<String>,
+        chain_id: u64,
+    ) -> Result<Self> {
+        ensure!(chain_id == 1, "Ethplorer API unsupported network");
+        Ok(Self {
+            client,
+            base: Url::try_from(BASE).unwrap(),
+            api_key: api_key.unwrap_or_else(|| FREE_API_KEY.to_owned()),
+            rate_limiter: None,
+        })
+    }
+
+    pub fn with_rate_limiter(&mut self, strategy: RateLimitingStrategy) -> &mut Self {
+        self.rate_limiter = Some(RateLimiter::from_strategy(strategy, "ethplorer".to_owned()));
+        self
+    }
+
+    async fn query_owners(&self, token: H160) -> Result<Vec<H160>> {
+        let mut url = self.base.join(&format!("{token:?}"))?;
+        // We technically only need one candidate, returning the top 2 in case there
+        // is a race condition and tokens have just been transferred out.
+        url.query_pairs_mut().append_pair("limit", "2");
+
+        tracing::debug!(%url, "querying Ethplorer");
+        // Don't log the API key!
+        url.query_pairs_mut().append_pair("apiKey", &self.api_key);
```

Comment:
Interesting idea. 👍 
We should probably do that for the paraswap partner ID as well and probably some other secrets.